src/cargo/ops/cargo_install.rs

```diff
@@ -182,16 +197,48 @@ fn install_one(
             }
         }
         src.update()?;
-        select_pkg(src, krate, vers, config, false, &mut |path| {
+        select_pkg(&mut src, krate, None, config, false, &mut |path| {
             path.read_packages()
         })?
     } else {
+        let mut source = map.load(source_id, &HashSet::new())?;
```

Comment:
This function is already pretty long and gnarly.  Can this new section of code somehow be moved out to another function?